### PR TITLE
fix(project): could delete users who are currently joining projects

### LINKED ISSUE #922: [Bug]: Deleting a user requires verifying whether the user is bound to the item
### ODC version

ODC 4.2.0

### OB version

-

### What happened?

If a user is both the administrator and dba of the project, after I delete this user, I can no longer operate the database in the project.

### What did you expect to happen?

If you want to delete the user, you should verify whether it is bound to the item.

### How can we reproduce it (as minimally and precisely as possible)?

1.create a user u1;
2.create a item p1,set u1 as administrator and dba;
3.Add a database to the project p1;
4.delete user u1;

### Anything else we need to know?

_No response_

### Cloud

_No response_

### PR DESCRIPTION
**Pull request by MarkPotato777**
#### What type of PR is this?

type-bug
module-project

#### What this PR does / why we need it:
When deleting users, we do not verify if they are project members which may cause the problem that some project doesn't have any OWNER after deletion, then nobody could assign a new OWNER to this project because only OWNER could assign project roles.
This PR adds a pre-delete-user hook to detect if the deleting user still joins some project, if yes then reject the operation.

#### Which issue(s) this PR fixes:

Fixes #922 

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java [hunk 1]
```diff
@@ -91,4 +108,20 @@ private void approvalFlowConfigUsageCheck(Long flowConfigId) {
         }
     }
 
+    private void projectReferenceCheck(Long userId, Long organizationId) {
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java:111]: As long as the current user is added to any of the project, they cannot be deleted. Would this be too difficult to delete a user.

File: server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java [hunk 2]
```diff
@@ -91,4 +108,20 @@ private void approvalFlowConfigUsageCheck(Long flowConfigId) {
...
+        Map<Long, Set<ResourceRoleName>> projectId2ResourceRoleNames =
+                resourceRoleService.getProjectId2ResourceRoleNames(userId);
+        if (projectId2ResourceRoleNames.size() == 0) {
+            return;
+        }
+        Map<Long, ProjectEntity> id2Project =
+                projectRepository.findAllByOrganizationId(organizationId).stream().filter(p -> p.getArchived() == false)
+                        .collect(Collectors.toMap(ProjectEntity::getId, p -> p));
+        throw new UnsupportedException(
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java:120]: Error messages should ideally be internationalized, see `ErrorCodes.CannotOperateDueReference`

**Review by smallsheeeep — CHANGES_REQUESTED** [review]

File: server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java (hunk 1 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java:111]: I changed the condition, now only if the user is the owner or dba then it cannot be deleted.

File: server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java (hunk 2 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java]: okay, done~

File: server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java [hunk 3]
```diff
@@ -91,4 +108,31 @@ private void approvalFlowConfigUsageCheck(Long flowConfigId) {
...
+                projectRepository.findAllByOrganizationId(organizationId).stream().filter(p -> p.getArchived() == false)
+                        .collect(Collectors.toMap(ProjectEntity::getId, p -> p));
+        String names = projectId2ResourceRoleNames.keySet().stream().map(id -> id2Project.get(id).getName())
+                .collect(Collectors.joining(", "));
+
+        String errorMessage = String.format(
+                "User id=%s cannot be deleted because it has been referenced to following project: {%s}", userId,
+                names);
+
+        throw new UnsupportedException(ErrorCodes.CannotOperateDueReference,
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java:130]: BadRequestException will be better？
UnsupportedException means that a feature is not supported yet

Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/config/HookConfiguration.java]: done

**Review by smallsheeeep — APPROVED** [review]

**Pull request merged**